electricitylci/combinator.py

```diff
@@ -0,0 +1,395 @@
+
+    expand_fuel_df = fuel_df["flowdict"].apply(pd.Series)
+    fuel_df.drop(columns=["flowdict"], inplace=True)
+
+    fuel_df["Compartment"] = "input"
+    fuel_df["FlowName"] = expand_fuel_df["q_reference_name"]
+    fuel_df["stage_code"] = upstream_reduced["stage_code"]
+    fuel_df["FlowAmount"] = upstream_reduced["quantity"]
+    fuel_df["FlowUUID"] = expand_fuel_df["q_reference_id"]
+    fuel_df["Unit"] = expand_fuel_df["q_reference_unit"]
+    fuel_df["eGRID_ID"] = upstream_df["plant_id"]
+    fuel_df["FacilityID"] = upstream_df["plant_id"]
+    fuel_df["FuelCategory"] = upstream_df["FuelCategory"]
+    fuel_df["Year"] = upstream_df["Year"]
+    merge_cols = [
+        "Age",
+        "Balancing Authority Code",
+        "Balancing Authority Name",
+        "Electricity",
+        #        "FRS_ID",
+        "NERC",
+        "Subregion",
+    ]
+    fuel_df.drop(columns=merge_cols, inplace=True)
+    gen_df_reduced = gen_df[merge_cols + ["eGRID_ID"]].drop_duplicates(
+        subset=["eGRID_ID"]
+    )
+
+    fuel_df = fuel_df.merge(
+        right=gen_df_reduced,
+        left_on="eGRID_ID",
+        right_on="eGRID_ID",
+        how="left",
+    )
+    fuel_df.dropna(subset=["Electricity"], inplace=True)
+    fuel_df["Source"] = "eia"
+    fuel_df = add_temporal_correlation_score(fuel_df)
+    fuel_df["DataCollection"] = 5
+    fuel_df["GeographicalCorrelation"] = 1
+    fuel_df["TechnologicalCorrelation"] = 1
+    fuel_df["ReliabilityScore"] = 1
+    fuel_df["ElementaryFlowPrimeContext"] = "input"
+    fuel_cat_key = (
+        gen_df[["FacilityID", "FuelCategory"]]
+        .drop_duplicates(subset="FacilityID")
+        .set_index("FacilityID")
+    )
+    fuel_df["FuelCategory"] = fuel_df["FacilityID"].map(
+        fuel_cat_key["FuelCategory"]
+    )
+    gen_plus_up_df = pd.concat([gen_df, fuel_df], ignore_index=True)
+    gen_plus_up_df = fill_nans(gen_plus_up_df)
+    return gen_plus_up_df
+
+
+if __name__ == "__main__":
+    import electricitylci.coal_upstream as coal
+    import electricitylci.natural_gas_upstream as ng
+    import electricitylci.petroleum_upstream as petro
+    import electricitylci.geothermal as geo
+    import electricitylci.solar_upstream as solar
+    import electricitylci.wind_upstream as wind
+    import electricitylci.nuclear_upstream as nuke
+
+    coal_df = coal.generate_upstream_coal(2016)
+    ng_df = ng.generate_upstream_ng(2016)
+    petro_df = petro.generate_petroleum_upstream(2016)
+    geo_df = geo.generate_upstream_geo(2016)
+    solar_df = solar.generate_upstream_solar(2016)
+    wind_df = wind.generate_upstream_wind(2016)
+    nuke_df = nuke.generate_upstream_nuc(2016)
+    upstream_df = concat_map_upstream_databases(
+        coal_df, ng_df, petro_df, geo_df, nuke_df, solar_df, wind_df
+    )
+    plant_df = altg.create_generation_process_df()
+    plant_df["stage_code"] = "Power plant"
+    combined_df = concat_clean_upstream_and_plant(plant_df, upstream_df)
```

Comment:
Getting another keyerror here, from the `fill_nans` function. Somehow the column `State` isn't included? 

```
INFO:ampd_plant_emissions.py:generate_plant_emissions:Summing eia923 fuel generation
INFO:ampd_plant_emissions.py:generate_plant_emissions:Summing eia923 boiler data
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating co2, ch4, n2o from gen fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating so2 emissions from gen fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating nox emissions from gen fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating co2, ch4, n2o emissions from boiler
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating so2 emissions from boiler fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating nox emissions from boiler fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Choosing emission sources
INFO:ampd_plant_emissions.py:generate_plant_emissions:Melting and mapping
WARNING:combinator.py:concat_clean_upstream_and_plant:Error deleting column Net Generation (MWh)
Traceback (most recent call last):
  File "combinator.py", line 389, in <module>
    combined_df = concat_clean_upstream_and_plant(plant_df, upstream_df)
  File "combinator.py", line 265, in concat_clean_upstream_and_plant
    combined_df = fill_nans(combined_df)
  File "combinator.py", line 77, in fill_nans
    df[[key_column] + target_columns]
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/frame.py", line 2934, in __getitem__
    raise_missing=True)
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/indexing.py", line 1354, in _convert_to_indexer
    return self._get_listlike_indexer(obj, axis, **kwargs)[1]
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/indexing.py", line 1161, in _get_listlike_indexer
    raise_missing=raise_missing)
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/indexing.py", line 1252, in _validate_read_indexer
    raise KeyError("{} not in index".format(not_found))
KeyError: "['State'] not in index"
```

Reply:
Not at all sure how this happened - would guess that somehow, there ended up being a "State_x" or "State_y". Anyhow, added a check to ensure column is in the df to avoid the keyerror. If state is missing, it will make a new column and then use the existing code to populate nans.

Pushed fix.